### PR TITLE
♻️ Refactor and Upgrade

### PR DESCRIPTION
**Pull request by jacobwoffenden**
This pull request:

- Updates `govuk_tech_docs` to `3.5.0`
  - This is not a "release" on [alphagov/tech-docs-gem](https://github.com/alphagov/tech-docs-gem) but is a tag 
- Updates Ruby to `3.2.3`
- Refactor container
- Refactor workflows
- Release to GitHub Container Registry
- Adds a testing workflow to ensure a basic tech-docs can be built

TODO:

- [x] Pin `govuk_tech_docs` to `3.4.0`
- [x] Update README
- [x] ~Local development of tech-docs via container on systems that don't use ID of `1000`~
- [x] Instructions for updating `Gemfile` and `Gemfile.lock`

Signed-off-by: Jacob Woffenden <jacob.woffenden@digital.justice.gov.uk>

### CODE REVIEW

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        run: |
+          docker build --file Dockerfile --tag ghcr.io/ministryofjustice/tech-docs-github-pages-publisher:latest .
+
+      - name: Test
+        id: test
+        shell: bash
+        run: |
+          docker run --rm \
+            --volume ${GITHUB_WORKSPACE}/test:/publisher \
+            ghcr.io/ministryofjustice/tech-docs-github-pages-publisher:latest /usr/local/bin/td-package
```
Comment (jasonBirchall) [inline .github/workflows/build-test.yml:34]: Love it ❤️

File: .github/workflows/dependency-review.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+---
+name: Dependency Review
```
Comment (jasonBirchall) [inline .github/workflows/dependency-review.yml:2]: We have a convention regarding our workflow names. It isn't well-published, so I can understand why this was missed. 

https://github.com/ministryofjustice/operations-engineering?tab=readme-ov-file#naming-standards-for-workflow-files

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        id: install_cosign
+        uses: sigstore/cosign-installer@e1523de7571e31dbe865fd2e80c5c7c23ae71eb4 # v3.4.0
+
+      - name: Log in to GitHub Container Registry
+        id: login_ghcr
+        uses: docker/login-action@343f7c4344506bcbf9b4de18042ae17996df046d # v3.0.0
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
Comment (jasonBirchall) [inline .github/workflows/release.yml:34]: I don't see the requirement to change container registries here. Can you explain your thinking?

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        uses: docker/build-push-action@4a13e500e55cf31b7a5d59a38ab2040ab0f42f56 # v5.1.0
+        with:
+          push: true
+          tags: ghcr.io/ministryofjustice/tech-docs-github-pages-publisher:${{ github.ref_name }}
+
+      - name: Sign
+        id: sign
+        shell: bash
+        run: |
+          cosign sign --yes ghcr.io/ministryofjustice/tech-docs-github-pages-publisher@${{ steps.build_and_push.outputs.digest }}
```
Comment (jasonBirchall) [inline .github/workflows/release.yml:47]: I like it, it makes sense, but are consumers going to check/care. Again it comes down to requirement.

File: Dockerfile [hunk 5]
```diff
@@ -1,29 +1,42 @@
-# Do not update to Ruby 3 until Gemfile dependencies are fixed
-FROM ruby:2.7.6-alpine3.15
+FROM public.ecr.aws/docker/library/ruby:3.2.3-alpine3.19
 
-# These are needed to support building native extensions during
-# bundle install step
-RUN apk --update add --virtual build_deps build-base git
+LABEL org.opencontainers.image.vendor="Ministry of Justice" \
+      org.opencontainers.image.authors="Operations Engineering" \
```
Comment (jasonBirchall) [inline Dockerfile:4]: This is not needed for a myriad of reasons. I don't particularly respect Dockerhub labels, but I'm happy for the others to stay, but authors must go.

File: Dockerfile [hunk 6]
```diff
@@ -1,29 +1,42 @@
...
+      org.opencontainers.image.url="https://github.com/ministryofjustice/tech-docs-github-pages-publisher"
 
-RUN addgroup -g 1000 -S appgroup \
-  && adduser -u 1000 --system appuser \
-  && adduser appuser appgroup \
-  && gem install bundler -v 2.4.22\
-  && bundle config
+ENV PUBLISHER_DIRECTORY="/publisher" \
+    BUNDLER_VERSION="2.5.6" \
+    LYCHEE_VERSION="0.14.3"
```
Comment (jasonBirchall) [inline Dockerfile:11]: I'm unconvinced of the benefit of installing lychee in the container. You can easily make it a step in a GitHub Action and avoid having to maintain the version at source i.e. https://github.com/ministryofjustice/operations-engineering-user-guide/blob/main/.github/workflows/build.yml#L37-L46

File: Dockerfile [hunk 7]
```diff
@@ -1,29 +1,42 @@
...
-# Required at runtime by middleman
-RUN apk add --no-cache nodejs
 
-WORKDIR /app
+# Install dependencies
+RUN apk --update-cache --no-cache add \
+      build-base \
+      curl \
+      git \
+      nodejs
```
Comment (jasonBirchall) [inline Dockerfile:19]: Without Lychee we don't need the added guff (git, curl etc)

File: LICENSE [hunk 8]
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2020-2022 Crown Copyright (Ministry of Justice)
+Copyright (c) 2020-2024 Crown Copyright (Ministry of Justice)
```
Comment (jasonBirchall) [inline LICENSE:3]: ❤️

File: README.md [hunk 9]
```diff
@@ -2,21 +2,23 @@
...
 
-This repository creates a Docker image and publishes the image to DockerHub.
+This repository publishes a utility container for previewing and packaging [GOV.UK's Technical Documentation Template](https://github.com/alphagov/tech-docs-template).
 
-The Docker image is pulled by other repository CI/CD, uses the container with their source code and call the scripts with the container to create documentation websites that meet government technical documents. The data created, ie html website, is then pushed to GH Pages by the repository CI/CD.
+## ⚠️ Breaking Changes
```
Comment (jasonBirchall) [inline README.md:7]: Much nicer 👍

File: README.md [hunk 10]
```diff
@@ -2,21 +2,23 @@
...
-- [check-url-links.sh](scripts/check-url-links.sh) - compiles the source code, places the HTML into the ./docs folder, and tests **internal and external URL links** using htmlproofer.
+The testing framework is now [Lychee.](https://github.com/lycheeverse/lychee)
 
-This image is used by the [MoJ Template Documentation Site](https://github.com/ministryofjustice/template-documentation-site) repository for MOJ technical documentation that gets published to GitHub Pages.
+The scripts have moved:
 
-## Breaking Change in v3
+- `/scripts/deploy.sh` is now `src/usr/local/bin/td-package`
+
+- `/scripts/preview.sh` is now `src/usr/local/bin/td-preview`
```
Comment (jasonBirchall) [inline README.md:19]: I'm not convinced this is an improvement on the current directory path. Why not just: `bin/preview`

File: src/opt/publisher/Gemfile.lock [hunk 11]
```diff
@@ -175,26 +170,23 @@ GEM
...
 
 PLATFORMS
-  ruby
+  aarch64-linux
+  arm-linux
+  arm64-darwin
+  x86-linux
+  x86_64-darwin
+  x86_64-linux
+  x86_64-linux-musl
```
Comment (jasonBirchall) [inline src/opt/publisher/Gemfile.lock:186]: Same question about requirements. It's great that we have options, but why?

File: src/opt/publisher/Gemfile.lock [hunk 12]
```diff
@@ -175,26 +170,23 @@ GEM
...
 
 DEPENDENCIES
-  govuk_tech_docs
-  haml (= 5.2.2)
-  html-proofer (= 4.4.3)
-
-RUBY VERSION
-   ruby 2.7.6p219
+  govuk_tech_docs (= 3.4.0)
```
Comment (jasonBirchall) [inline src/opt/publisher/Gemfile.lock:189]: I may have misunderstood what this means, but 3.3.1 is the latest version:
https://github.com/alphagov/tech-docs-gem/releases/tag/v3.3.1

File: .github/workflows/cicd-push-docker-image.yml [hunk 13]
```diff
@@ -1,42 +0,0 @@
...
-
-      - name: Report failure to Slack
-        if: always()
-        uses: ravsamhq/notify-slack-action@v2
-        with:
-          status: ${{ job.status }}
-          notify_when: "failure"
-          notification_title: "Failed GitHub Action Run"
-        env:
-          SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
```
Comment (jasonBirchall) [inline .github/workflows/cicd-push-docker-image.yml:42]: This is actually fairly improtant.

File: Dockerfile (hunk 7 above)
Comment (jacobwoffenden) [reply Dockerfile:19]: Lychee removed 👋

File: src/opt/publisher/Gemfile.lock (hunk 12 above)
Comment (jacobwoffenden) [reply src/opt/publisher/Gemfile.lock:189]: 3.3.1 is the latest "release" but not the latest tag

File: src/opt/publisher/Gemfile.lock (hunk 11 above)
Comment (jacobwoffenden) [reply src/opt/publisher/Gemfile.lock:186]: This is generated programatically

**Review by jasonBirchall — APPROVED** [review]

**Review by jasonBirchall — APPROVED** [review]

**Pull request merged**